packages/core/src/messagingProtocols/mls/E2eIdentityService/E2eIdentityService.ts

```diff
@@ -0,0 +1,321 @@
+        }
+        AcmeStorage.storeInitialData({discoveryUrl, user, clientId});
+        await E2eIdentityService.instance.init({clientId, discoveryUrl, user});
+      }
+    }
+    return E2eIdentityService.instance;
+  }
+
+  public async getNewCertificate(): Promise<boolean> {
+    // Step 0: Check if we have a handle in local storage
+    // If we don't have a handle, we need to start a new OAuth flow
+    this.isInProgress = AcmeStorage.hasHandle();
+
+    if (this.isInProgress) {
+      try {
+        return await this.continueOAuthFlow();
+      } catch (error) {
+        return this.exitWithError('Error while trying to continue OAuth flow with error:', error);
+      }
+    } else {
+      try {
+        return await this.startNewOAuthFlow();
+      } catch (error) {
+        return this.exitWithError('Error while trying to start OAuth flow with error:', error);
+      }
+    }
+  }
+
+  // ############ Internal Functions ############
+
+  private exitWithError(message: string, error?: unknown) {
+    this.logger.error(message, error);
+    return false;
+  }
+
+  private async init(params: Required<Pick<InitParams, 'user' | 'clientId' | 'discoveryUrl'>>): Promise<void> {
+    const {user, clientId, discoveryUrl} = params;
+    if (!user || !clientId) {
+      this.logger.error('user and clientId are required to initialize E2eIdentityService');
+      throw new Error();
+    }
+    this.acmeService = new AcmeService(discoveryUrl);
+    this.user = user;
+    this.clientId = clientId;
+    this.e2eClientId = getE2eClientId(this.user, this.clientId);
+    this.identity = await this.coreCryptoClient.e2eiNewEnrollment(
+      this.e2eClientId,
+      this.user.displayName,
+      this.user.handle,
+      this.expiryDays,
+      Ciphersuite.MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519,
+    );
+    this.isInitialized = true;
+  }
+
+  private async getDirectory(identity: WireE2eIdentity, connection: AcmeService): Promise<AcmeDirectory | undefined> {
+    try {
+      const directory = await connection.getDirectory();
+
+      if (directory) {
+        const parsedDirectory = identity.directoryResponse(directory);
+        return parsedDirectory;
+      }
+    } catch (e) {
+      throw new Error('Error while trying to receive a directory', {cause: e});
+    }
+    return undefined;
+  }
+
+  private async getInitialNonce(directory: AcmeDirectory, connection: AcmeService): Promise<string> {
+    try {
+      const nonce = await connection.getInitialNonce(directory.newNonce);
+      if (nonce) {
+        return nonce;
+      }
+      throw new Error('No initial-nonce received');
+    } catch (e) {
+      throw new Error('Error while trying to receive a nonce', {cause: e});
+    }
+  }
+
+  private getOidcService(challenge: AcmeChallenge): OIDCService {
+    const oidcService = new OIDCService({
+      audience: '338888153072-ktbh66pv3mr0ua0dn64sphgimeo0p7ss.apps.googleusercontent.com',
```

Comment:
These will be configured via env variables, right?

Reply:
Yes, as soon as we have a test environment for OAuth. For now, its just using google oauth. 
But in the end, every customer will decide their environment themselves.